lib/configgin.rb

```diff
@@ -58,73 +57,86 @@ def generate_jobs(job_configs, templates)
+  end
+
+  # Write exported properties to secret and potentially restart affected pods.
+  def export_job_properties(jobs)
+    # co-located containers don't get to export properties
+    return unless instance_group == ENV["KUBERNETES_CONTAINER_NAME"]
+    # jobs don't export properties
+    return unless self_pod['metadata']['ownerReferences'][0]['kind'] == "StatefulSet"
 
+    sts = kube_client_stateful_set.get_stateful_set(instance_group, kube_namespace)
+
+    # Make sure the secret attached to the stateful set exists.
+    # XXX This should probably be done by fissile via the helm chart.
     secret = Kubeclient::Resource.new
-    secret.metadata = {}
-    # Prefixing with pod.metadata.name is purely for human convenience/debugging.
-    secret.metadata.name = "#{pod.metadata.name}-#{pod.metadata.uid}"
-    secret.metadata.namespace = kube_namespace
-
-    # Make sure the secret gets removed when the pod is deleted.
-    secret.metadata.ownerReferences = [
-      {
-        apiVersion: pod.apiVersion,
-        blockOwnerDeletion: false,
-        controller: false,
-        kind: pod.kind,
-        name: pod.metadata.name,
-        uid: pod.metadata.uid,
-      }
-    ]
+    secret.metadata = {
+      name: sts.metadata.name,
+      namespace: kube_namespace,
+      ownerReferences:  [
+        {
+          apiVersion: sts.apiVersion,
+          blockOwnerDeletion: false,
+          controller: false,
+          kind: sts.kind,
+          name: sts.metadata.name,
+          uid: sts.metadata.uid,
+        }
+      ]
+    }
+    begin
+      kube_client.create_secret(secret)
+    rescue
+    end
+    secret = kube_client.get_secret(instance_group, kube_namespace)
+    secret.data ||= {}
+
+    version_tag = ENV["CONFIGGIN_VERSION_TAG"]
+    new_tag = !secret.data.has_key?(version_tag)
+    secret.data = {version_tag => ""} if new_tag # make sure old properties are deleted during upgrade
 
-    secret.data = {}
     digests = {}
     jobs.each do |name, job|
       digests[name] = property_digest(job.exported_properties)
       secret.data["skiff-exported-properties-#{name}"] = Base64.encode64(job.exported_properties.to_json)
-      secret.data["skiff-exported-digest-#{name}"] = Base64.encode64(digests[name])
-    end
 
-    # Only the main container gets to export properties; colocated sidecars don't.
-    if instance_group == ENV["KUBERNETES_CONTAINER_NAME"]
-      begin
-        kube_client.delete_secret(secret.metadata.name, kube_namespace)
-      rescue
+      encoded_digest = Base64.encode64(digests[name])
+
+      # Record initial digest values whenever the tag changes, in which case the pod startup
+      # order is already controlled by the "CONFIGGIN_IMPORT_#{role}" references to the new
+      # tags in the corresponding secrets. There is no annotation when importing this set of
+      # initial values because the helm chart doesn't include any annotations, and we don't
+      # want to trigger a pod restart by adding them.
```

Comment:
Okay, trying to rephrase this just to see if I understand this correctly (no need to take my wording):

> In order to ensure that we restart pods when their imported properties are changed, we set annotations on the importing StatefulSet containing digests of the imported properties.

> On initial deployment, the StatefulSets do not have those annotations (due to the helm deployment); to avoid unnecessarily restarting those initial pods, we explicitly do _not_ set annotations for the initial values (when we have a new version tag).  In that case, pod startup ordering is controlled by the `CONFIGGIN_IMPORT_#{role}` environment variable references to the secret.

> In order to keep track of whether this is a new version tag, we place the version tag in the generated secret.

Reply:
> In order to keep track of whether this is a new version tag, we place the version tag in the generated secret.

We place the tag into the secret so fissile can create the `CONFIGGIN_IMPORT_role` variables that reference them, so kube will delay container creation until the secret exists.

And we use the version tag because during updates (or secrets rotation) the old tag already exists, but we don't want that version of the secrets, we have to wait until they are replace by the new ones.